File: ios/Classes/KlaviyoFlutterPlugin.swift
import Flutter
import KlaviyoSwift
import UIKit

// swiftlint:disable identifier_name type_body_length file_length

/// A class that receives and handles calls from Flutter to complete the payment.
public class KlaviyoFlutterPlugin: NSObject, FlutterPlugin,
    UNUserNotificationCenterDelegate {
    private static let methodChannelName = "com.rightbite.denisr/klaviyo"

    private let METHOD_UPDATE_PROFILE = "updateProfile"
    private let METHOD_INITIALIZE = "initialize"
    private let METHOD_SEND_TOKEN = "sendTokenToKlaviyo"
    private let METHOD_SET_BADGE_COUNT = "setBadgeCount"
    private let METHOD_LOG_EVENT = "logEvent"
    private let METHOD_HANDLE_PUSH = "handlePush"
    private let METHOD_SET_EXTERNAL_ID = "setExternalId"
    private let METHOD_GET_EXTERNAL_ID = "getExternalId"
    private let METHOD_RESET_PROFILE = "resetProfile"
    private let METHOD_SET_EMAIL = "setEmail"
    private let METHOD_GET_EMAIL = "getEmail"
    private let METHOD_SET_PHONE_NUMBER = "setPhoneNumber"
    private let METHOD_GET_PHONE_NUMBER = "getPhoneNumber"
    private let METHOD_SET_FIRST_NAME = "setFirstName"
    private let METHOD_SET_LAST_NAME = "setLastName"
    private let METHOD_SET_ORGANIZATION = "setOrganization"
    private let METHOD_SET_TITLE = "setTitle"
    private let METHOD_SET_IMAGE = "setImage"
    private let METHOD_SET_ADDRESS1 = "setAddress1"
    private let METHOD_SET_ADDRESS2 = "setAddress2"
    private let METHOD_SET_CITY = "setCity"
    private let METHOD_SET_COUNTRY = "setCountry"
    private let METHOD_SET_LATITUDE = "setLatitude"
    private let METHOD_SET_LONGITUDE = "setLongitude"
    private let METHOD_SET_REGION = "setRegion"
    private let METHOD_SET_ZIP = "setZip"
    private let METHOD_SET_TIMEZONE = "setTimezone"
    private let METHOD_SET_CUSTOM_ATTRIBUTE = "setCustomAttribute"

    private let klaviyo = KlaviyoSDK()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let channel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: messenger
        )
        let instance = KlaviyoFlutterPlugin()

        if #available(OSX 10.14, *) {
            let center = UNUserNotificationCenter.current()
            center.delegate = instance
        }

        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // below method will be called when the user interacts with the push notification
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {

        // If this notification is Klaviyo's notification we'll handle it
        // else pass it on to the next push notification service to which it may belong
        let handled = KlaviyoSDK().handle(
            notificationResponse: response,
            withCompletionHandler: completionHandler
        )
        if !handled {
            completionHandler()
        }
    }

    // below method is called when the app receives push notifications when the app is the foreground
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (
            UNNotificationPresentationOptions
        ) -> Void
    ) {
        var options: UNNotificationPresentationOptions = [.alert]
        if #available(iOS 14.0, *) {
            options = [.list, .banner]
        }
        completionHandler(options)
    }

    // swiftlint:disable function_body_length cyclomatic_complexity
    public func handle(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        func setProfileAttribute(
            key: Profile.ProfileKey,
            name: String,
            argumentKey: String
        ) {
            guard
                let arguments = call.arguments as? [String: Any],
                let stringValue = arguments[argumentKey] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "\(name) must be a non-null String",
                        details: nil
                    )
                )
            }
            klaviyo.set(profileAttribute: key, value: stringValue)
            result("\(name) updated")
        }

        func setNumericProfileAttribute(
            call: FlutterMethodCall,
            key: Profile.ProfileKey,
            name: String,
            argumentKey: String,
            result: FlutterResult
        ) {
            guard
                let arguments = call.arguments as? [String: Any],
                let numericValue = extractDouble(arguments[argumentKey])
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "\(name) must be a numeric value",
                        details: nil
                    )
                )
            }
            klaviyo.set(profileAttribute: key, value: numericValue)
            result("\(name) updated")
        }

        switch call.method {
        case METHOD_INITIALIZE:
            guard
                let arguments = call.arguments as? [String: Any],
                let apiKey = arguments["apiKey"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "API key must be provided",
                        details: nil
                    )
                )
            }
            klaviyo.initialize(with: apiKey)
            result("Klaviyo initialized")

        case METHOD_SEND_TOKEN:
            guard
                let arguments = call.arguments as? [String: Any],
                let tokenData = arguments["token"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Token must be provided",
                        details: nil
                    )
                )
            }
            klaviyo.set(pushToken: Data(hexString: tokenData))
            result("Token sent to Klaviyo")

        case METHOD_SET_BADGE_COUNT:
            guard
                let arguments = call.arguments as? [String: Any],
                let count = arguments["count"] as? Int
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "count must be an Int",
                        details: nil
                    )
                )
            }
            DispatchQueue.main.async {
                self.klaviyo.setBadgeCount(count)
                result("Badge count set")
            }

        case METHOD_UPDATE_PROFILE:
            guard
                let arguments = call.arguments as? [String: Any]
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Profile must be provided",
                        details: nil
                    )
                )
            }
            // parsing location
            let address1 = arguments["address1"] as? String
            let address2 = arguments["address2"] as? String
            let city = arguments["city"] as? String
            let country = arguments["country"] as? String
            let region = arguments["region"] as? String
            let zip = arguments["zip"] as? String
            let timezone = arguments["timezone"] as? String
            let latitude = extractDouble(arguments["latitude"])
            let longitude = extractDouble(arguments["longitude"])

            let hasLocationData = [
                address1,
                address2,
                city,
                country,
                region,
                zip,
                timezone
            ].contains { value in value != nil } || latitude != nil || longitude != nil

            let location = hasLocationData
                ? Profile.Location(
                    address1: address1,
                    address2: address2,
                    city: city,
                    country: country,
                    latitude: latitude,
                    longitude: longitude,
                    region: region,
                    zip: zip,
                    timezone: timezone
                )
                : nil

            let profile = Profile(
                email: arguments["email"] as? String,
                phoneNumber: arguments["phone_number"] as? String,
                externalId: arguments["external_id"] as? String,
                firstName: arguments["first_name"] as? String,
                lastName: arguments["last_name"] as? String,
                organization: arguments["organization"] as? String,
                title: arguments["title"] as? String,
                image: arguments["image"] as? String,
                location: location,
                properties: arguments["properties"] as? [String: Any]
            )
            klaviyo.set(profile: profile)
            result("Profile updated")

        case METHOD_LOG_EVENT:
            guard
                let arguments = call.arguments as? [String: Any],
                let eventName = arguments["name"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Event name must be provided",
                        details: nil
                    )
                )
            }
            let event = Event(
                name: .customEvent(eventName),
                properties: arguments["metaData"] as? [String: Any]
            )

            klaviyo.create(event: event)
            result("Event: [\(event)] created")

        case METHOD_HANDLE_PUSH:
            guard
                let arguments = call.arguments as? [String: Any]
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Push message must be provided",
                        details: nil
                    )
                )
            }

            if let properties = arguments["message"] as? [String: Any],
                properties["_k"] != nil {
                klaviyo.create(
                    event: Event(
                        name: .customEvent("$opened_push"),
                        properties: properties
                    )
                )

                return result(true)
            }
            result(false)

        case METHOD_GET_EXTERNAL_ID:
            result(klaviyo.externalId)

        case METHOD_RESET_PROFILE:
            klaviyo.resetProfile()
            result(true)

        case METHOD_GET_EMAIL:
            result(klaviyo.email)

        case METHOD_GET_PHONE_NUMBER:
            result(klaviyo.phoneNumber)

        case METHOD_SET_EXTERNAL_ID:
            guard
                let arguments = call.arguments as? [String: Any],
                let externalId = arguments["id"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "External ID must be provided",
                        details: nil
                    )
                )
            }
            klaviyo.set(externalId: externalId)
            result("ID updated")

        case METHOD_SET_EMAIL:
            guard
                let arguments = call.arguments as? [String: Any],
                let email = arguments["email"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Email must be provided",
                        details: nil
                    )
                )
            }

            klaviyo.set(email: email)
            result("Email updated")

        case METHOD_SET_PHONE_NUMBER:
            guard
                let arguments = call.arguments as? [String: Any],
                let phoneNumber = arguments["phoneNumber"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Phone number must be provided",
                        details: nil
                    )
                )
            }

            klaviyo.set(phoneNumber: phoneNumber)
            result("Phone updated")

        case METHOD_SET_FIRST_NAME:
            setProfileAttribute(
                key: .firstName,
                name: "First name",
                argumentKey: "firstName"
            )

        case METHOD_SET_LAST_NAME:
            setProfileAttribute(
                key: .lastName,
                name: "Last name",
                argumentKey: "lastName"
            )

        case METHOD_SET_TITLE:
            setProfileAttribute(
                key: .title,
                name: "Title",
                argumentKey: "title"
            )

        case METHOD_SET_ORGANIZATION:
            setProfileAttribute(
                key: .organization,
                name: "Organization",
                argumentKey: "organization"
            )

        case METHOD_SET_IMAGE:
            setProfileAttribute(
                key: .image,
                name: "Image",
                argumentKey: "image"
            )

        case METHOD_SET_ADDRESS1:
            setProfileAttribute(
                key: .address1,
                name: "Address 1",
                argumentKey: "address"
            )

        case METHOD_SET_ADDRESS2:
            setProfileAttribute(
                key: .address2,
                name: "Address 2",
                argumentKey: "address"
            )

        case METHOD_SET_CITY:
            setProfileAttribute(key: .city, name: "City", argumentKey: "city")

        case METHOD_SET_COUNTRY:
            setProfileAttribute(
                key: .country,
                name: "Country",
                argumentKey: "country"
            )

        case METHOD_SET_LATITUDE:
            setNumericProfileAttribute(
                call: call,
                key: .latitude,
                name: "Latitude",
                argumentKey: "latitude",
                result: result
            )

        case METHOD_SET_LONGITUDE:
            setNumericProfileAttribute(
                call: call,
                key: .longitude,
                name: "Longitude",
                argumentKey: "longitude",
                result: result
            )

        case METHOD_SET_REGION:
            setProfileAttribute(
                key: .region,
                name: "Region",
                argumentKey: "region"
            )

        case METHOD_SET_ZIP:
            setProfileAttribute(key: .zip, name: "Zip", argumentKey: "zip")

        case METHOD_SET_TIMEZONE:
            // Klaviyo takes timezone from environment on iOS
            result("Success")

        case METHOD_SET_CUSTOM_ATTRIBUTE:
            guard
                let arguments = call.arguments as? [String: Any],
                let key = arguments["key"] as? String,
                let value = arguments["value"] as? String
            else {
                return result(
                    FlutterError(
                        code: "invalid_args",
                        message: "Method setCustomAttribute requires arguments {key: String, value: String}",
                        details: nil
                    )
                )
            }
            klaviyo.set(profileAttribute: .custom(customKey: key), value: value)
            result("Attribute \(key) updated")

        default:
            result(FlutterMethodNotImplemented)
        }
    }
    // swiftlint:enable function_body_length cyclomatic_complexity
}

extension Data {
    init(hexString: String) {
        self =
            hexString
            .dropFirst(hexString.hasPrefix("0x") ? 2 : 0)
            .compactMap { $0.hexDigitValue.map { UInt8($0) } }
            .reduce(
                into: (
                    data: Data(capacity: hexString.count / 2),
                    byte: nil as UInt8?
                )
            ) { partialResult, nibble in
                if let pendingNibble = partialResult.byte {
                    partialResult.data.append(pendingNibble + nibble)
                    partialResult.byte = nil
                } else {
                    partialResult.byte = nibble << 4
                }
            }.data
    }
}

// swiftlint:enable identifier_name type_body_length

private func extractDouble(_ rawValue: Any?) -> Double? {
    if let value = rawValue as? Double {
        return value
    }
    if let numberValue = rawValue as? NSNumber {
        return numberValue.doubleValue
    }
    if let stringValue = rawValue as? String {
        return Double(stringValue)
    }
    return nil
}

File: android/src/main/kotlin/com/rightbite/denisr/KlaviyoFlutterPlugin.kt
package com.rightbite.denisr

import android.content.Context
import androidx.annotation.VisibleForTesting
import com.klaviyo.analytics.Klaviyo
import com.klaviyo.analytics.model.Event
import com.klaviyo.analytics.model.EventKey
import com.klaviyo.analytics.model.EventMetric
import com.klaviyo.analytics.model.Profile
import com.klaviyo.analytics.model.ProfileKey
import io.flutter.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.io.Serializable

private const val METHOD_UPDATE_PROFILE = "updateProfile"
private const val METHOD_INITIALIZE = "initialize"
private const val METHOD_SEND_TOKEN = "sendTokenToKlaviyo"
private const val METHOD_LOG_EVENT = "logEvent"
private const val METHOD_HANDLE_PUSH = "handlePush"
private const val METHOD_SET_EXTERNAL_ID = "setExternalId"
private const val METHOD_GET_EXTERNAL_ID = "getExternalId"
private const val METHOD_RESET_PROFILE = "resetProfile"
private const val METHOD_SET_EMAIL = "setEmail"
private const val METHOD_GET_EMAIL = "getEmail"
private const val METHOD_SET_PHONE_NUMBER = "setPhoneNumber"
private const val METHOD_GET_PHONE_NUMBER = "getPhoneNumber"
private const val METHOD_SET_FIRST_NAME = "setFirstName"
private const val METHOD_SET_LAST_NAME = "setLastName"
private const val METHOD_SET_ORGANIZATION = "setOrganization"
private const val METHOD_SET_TITLE = "setTitle"
private const val METHOD_SET_IMAGE = "setImage"
private const val METHOD_SET_ADDRESS1 = "setAddress1"
private const val METHOD_SET_ADDRESS2 = "setAddress2"
private const val METHOD_SET_CITY = "setCity"
private const val METHOD_SET_COUNTRY = "setCountry"
private const val METHOD_SET_LATITUDE = "setLatitude"
private const val METHOD_SET_LONGITUDE = "setLongitude"
private const val METHOD_SET_REGION = "setRegion"
private const val METHOD_SET_ZIP = "setZip"
private const val METHOD_SET_TIMEZONE = "setTimezone"
private const val METHOD_SET_CUSTOM_ATTRIBUTE = "setCustomAttribute"
private const val METHOD_SET_BADGE_COUNT = "setBadgeCount"

private const val PROFILE_PROPERTIES_KEY = "properties"

// Maps standard profile keys to their Kotlin object suffix names.
// We need reflection because the underlying ProfileKey objects are marked `internal`.
private val PROFILE_KEY_CLASS_SUFFIX =
    mapOf(
        "external_id" to "EXTERNAL_ID",
        "email" to "EMAIL",
        "phone_number" to "PHONE_NUMBER",
        "first_name" to "FIRST_NAME",
        "last_name" to "LAST_NAME",
        "organization" to "ORGANIZATION",
        "title" to "TITLE",
        "image" to "IMAGE",
        "address1" to "ADDRESS1",
        "address2" to "ADDRESS2",
        "city" to "CITY",
        "country" to "COUNTRY",
        "region" to "REGION",
        "zip" to "ZIP",
        "timezone" to "TIMEZONE",
        "latitude" to "LATITUDE",
        "longitude" to "LONGITUDE",
    )

private val profileKeyCache = mutableMapOf<String, ProfileKey?>()

private const val TAG = "KlaviyoFlutterPlugin"

class KlaviyoFlutterPlugin :
    MethodCallHandler,
    FlutterPlugin {
    private var applicationContext: Context? = null
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        applicationContext = null
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        fun setProfileAttribute(
            key: ProfileKey,
            name: String,
            argumentKey: String,
        ) {
            try {
                val value: String =
                    call.argument<String>(argumentKey)
                        ?: return result.error("Bad Request", "$name should not be null", null)
                Klaviyo.setProfileAttribute(
                    propertyKey = key,
                    value = value,
                )
                logInfo("$name updated")
                return result.success("$name updated")
            } catch (e: Exception) {
                return result.error("Set profile attribute error", e.message, e)
            }
        }

        when (call.method) {
            METHOD_INITIALIZE -> {
                val apiKey = call.argument<String>("apiKey")
                Klaviyo.initialize(apiKey!!, applicationContext!!)
                logDebug("initialized apiKey: $apiKey")
                result.success("Klaviyo initialized")
            }

            METHOD_SEND_TOKEN -> {
                val pushToken = call.argument<String>("token")
                if (pushToken != null) {
                    Klaviyo.setPushToken(pushToken)

                    result.success("Token sent to Klaviyo")
                }
            }

            METHOD_UPDATE_PROFILE -> {
                try {
                    val profilePropertiesRaw =
                        call.arguments<Map<String, Any>?>()
                            ?: throw RuntimeException("Profile properties not exist")

                    val serializedProperties = convertMapToSeralizedMap(profilePropertiesRaw)
                    val profileAttributes = buildProfileAttributes(serializedProperties)

                    val profile = Profile(profileAttributes)

                    Klaviyo.setProfile(profile)
                    logDebug("Profile updated: ${Klaviyo.getExternalId()}, profileMap: $serializedProperties")

                    result.success("Profile updated")
                } catch (e: Exception) {
                    result.error("Profile update error", e.message, e)
                }
            }

            METHOD_LOG_EVENT -> {
                val eventName =
                    call.argument<String>("name")
                        ?: return result.error("Bad Request", "Event name should not be null", null)
                val metaDataRaw = call.argument<Map<String, Any>?>("metaData")

                val event = Event(EventMetric.CUSTOM(eventName))

                if (metaDataRaw != null) {
                    val metaData = convertMapToSeralizedMap(metaDataRaw)
                    metaData.forEach { (key, value) ->
                        event.setProperty(EventKey.CUSTOM(key), value = value)
                    }
                }

                Klaviyo.createEvent(event)

                val metadataKeys = metaDataRaw?.keys?.joinToString(prefix = "[", postfix = "]")
                logDebug("Event created for '$eventName' with metadata keys $metadataKeys")
                result.success("Event[$eventName] created")
            }

            METHOD_HANDLE_PUSH -> {
                val metaData =
                    call.argument<HashMap<String, String>>("message") ?: emptyMap<String, String>()

                if (isKlaviyoPush(metaData)) {
                    val event =
                        Event(
                            EventMetric.CUSTOM("\$opened_push"),
                            metaData.mapKeys {
                                EventKey.CUSTOM(it.key)
                            },
                        )
                    return try {
                        Klaviyo.getPushToken()?.let { event[EventKey.CUSTOM("push_token")] = it }

                        Klaviyo.createEvent(event)
                        result.success(true)
                    } catch (e: Exception) {
                        logError("Failed handle push metaData:$metaData", e)
                        result.error("Failed handle push metaData", e.message, null)
                    }
                } else {
                    return result.success(false)
                }
            }

            METHOD_SET_EXTERNAL_ID -> {
                val id: String =
                    call.argument<String>("id")
                        ?: return result.error("Bad Request", "ID should not be null", null)
                Klaviyo.setExternalId(id)
                return result.success(null)
            }

            METHOD_GET_EXTERNAL_ID -> result.success(Klaviyo.getExternalId())

            METHOD_RESET_PROFILE -> {
                Klaviyo.resetProfile()
                result.success(true)
            }

            METHOD_GET_EMAIL -> result.success(Klaviyo.getEmail())
            METHOD_GET_PHONE_NUMBER -> result.success(Klaviyo.getPhoneNumber())

            METHOD_SET_EMAIL -> {
                call.argument<String>("email")?.let { newEmail ->
                    Klaviyo.setEmail(newEmail)
                    result.success("Email updated")
                }
            }

            METHOD_SET_PHONE_NUMBER -> {
                call.argument<String>("phoneNumber")?.let { newPhone ->
                    Klaviyo.setPhoneNumber(newPhone)
                    result.success("Phone number updated")
                }
            }

            METHOD_SET_FIRST_NAME -> {
                setProfileAttribute(ProfileKey.FIRST_NAME, "First name", "firstName")
            }

            METHOD_SET_LAST_NAME -> {
                setProfileAttribute(ProfileKey.LAST_NAME, "Last name", "lastName")
            }

            METHOD_SET_ORGANIZATION -> {
                setProfileAttribute(ProfileKey.ORGANIZATION, "Organization", "organization")
            }

            METHOD_SET_TITLE -> {
                setProfileAttribute(ProfileKey.TITLE, "Title", "title")
            }

            METHOD_SET_IMAGE -> {
                setProfileAttribute(ProfileKey.IMAGE, "Image", "image")
            }

            METHOD_SET_ADDRESS1 -> {
                setProfileAttribute(ProfileKey.ADDRESS1, "Address 1", "address")
            }

            METHOD_SET_ADDRESS2 -> {
                setProfileAttribute(ProfileKey.ADDRESS2, "Address 2", "address")
            }

            METHOD_SET_CITY -> {
                setProfileAttribute(ProfileKey.CITY, "City", "city")
            }

            METHOD_SET_COUNTRY -> {
                setProfileAttribute(ProfileKey.COUNTRY, "Country", "country")
            }

            METHOD_SET_LATITUDE -> {
                val latitude: Double =
                    call.argument<Double>("latitude")
                        ?: return result.error("Bad Request", "Latitude should not be null", null)
                Klaviyo.setProfileAttribute(ProfileKey.LATITUDE, latitude)
                logInfo("Latitude updated")
                result.success("Latitude updated")
            }

            METHOD_SET_LONGITUDE -> {
                val longitude: Double =
                    call.argument<Double>("longitude")
                        ?: return result.error("Bad Request", "Longitude should not be null", null)
                Klaviyo.setProfileAttribute(ProfileKey.LONGITUDE, longitude)
                logInfo("Longitude updated")
                result.success("Longitude updated")
            }

            METHOD_SET_REGION -> {
                setProfileAttribute(ProfileKey.REGION, "Region", "region")
            }

            METHOD_SET_ZIP -> {
                setProfileAttribute(ProfileKey.ZIP, "Zip", "zip")
            }

            METHOD_SET_TIMEZONE -> {
                setProfileAttribute(ProfileKey.TIMEZONE, "Timezone", "timezone")
            }

            METHOD_SET_CUSTOM_ATTRIBUTE -> {
                val key: String =
                    call.argument<String>("key")
                        ?: return result.error("Bad Request", "Key must not be null", null)
                val value: String =
                    call.argument<String>("value")
                        ?: return result.error("Bad Request", "Value must not be null", null)
                Klaviyo.setProfileAttribute(propertyKey = ProfileKey.CUSTOM(key), value)
                return result.success("Attribute '$key' updated")
            }

            METHOD_SET_BADGE_COUNT -> {
                val count: Int =
                    call.argument<Int>("count")
                        ?: return result.error("Bad Request", "count must not be null", null)
                logInfo("setBadgeCount($count) is not supported on Android; ignoring")
                return result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun isKlaviyoPush(payload: Map<String, String>) = payload.containsKey("_k")

    private fun logDebug(message: String) = runCatching { Log.d(TAG, message) }

    private fun logInfo(message: String) = runCatching { Log.i(TAG, message) }

    private fun logError(
        message: String,
        throwable: Throwable? = null,
    ) = runCatching {
        if (throwable != null) {
            Log.e(TAG, message, throwable)
        } else {
            Log.e(TAG, message)
        }
    }

    companion object {
        private const val CHANNEL_NAME = "com.rightbite.denisr/klaviyo"
    }
}

@VisibleForTesting
internal fun buildProfileAttributes(profileProperties: Map<String, Serializable>): Map<ProfileKey, Serializable> {
    val attributes = mutableMapOf<ProfileKey, Serializable>()
    val processedKeys = mutableSetOf<String>()

    // Process standard profile keys
    PROFILE_KEY_CLASS_SUFFIX.forEach { (rawKey, _) ->
        val value = profileProperties[rawKey] ?: return@forEach
        val profileKey = resolveStandardProfileKey(rawKey)
        if (profileKey != null) {
            attributes[profileKey] = value
            processedKeys.add(rawKey)
        }
    }

    // Process nested custom properties
    val customProperties = profileProperties[PROFILE_PROPERTIES_KEY] as? Map<*, *>
    customProperties?.forEach { (rawKey, rawValue) ->
        val key = rawKey as? String ?: return@forEach
        val value = rawValue as? Serializable ?: return@forEach
        attributes[ProfileKey.CUSTOM(key)] = value
    }

    // Process any remaining properties as custom attributes
    profileProperties.forEach { (key, value) ->
        if (!processedKeys.contains(key) && key != PROFILE_PROPERTIES_KEY) {
            attributes[ProfileKey.CUSTOM(key)] = value
        }
    }

    return attributes
}

@VisibleForTesting
internal fun resolveStandardProfileKey(rawKey: String): ProfileKey? =
    profileKeyCache.getOrPut(rawKey) {
        PROFILE_KEY_CLASS_SUFFIX[rawKey]?.let { suffix ->
            runCatching {
                val className = "com.klaviyo.analytics.model.ProfileKey$$suffix"
                val clazz = Class.forName(className)
                val instanceField = clazz.getField("INSTANCE")
                instanceField.get(null) as? ProfileKey
            }.getOrNull()
        }
    }

private fun convertMapToSeralizedMap(map: Map<String, Any?>): Map<String, Serializable> {
    val convertedMap = mutableMapOf<String, Serializable>()

    for ((key, value) in map) {
        // Only include non-null serializable values
        // Null values are skipped (cannot clear fields via this method)
        if (value is Serializable) {
            convertedMap[key] = value
        }
    }

    return convertedMap
}
